File: jlox/src/Scanner.java
package src;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import static src.TokenType.*;

public class Scanner {
    private final String source;
    private final List<Token> tokens = new ArrayList<>();

    private int start = 0;   // Points to first char in the current lexeme.
    private int current = 0; // Points to the current char in the source.
    private int line = 1;

    private static final Map<String, TokenType> keywords = new HashMap<>();
    static {
        keywords.put("and",      AND);
        keywords.put("break",    BREAK);
        keywords.put("class",    CLASS);
        keywords.put("continue", CONTINUE);
        keywords.put("do",       DO);
        keywords.put("else",     ELSE);
        keywords.put("false",    FALSE);
        keywords.put("for",      FOR);
        keywords.put("fun",      FUN);
        keywords.put("if",       IF);
        keywords.put("nil",      NIL);
        keywords.put("or",       OR);
        keywords.put("print",    PRINT);
        keywords.put("println",  PRINTLN);
        keywords.put("return",   RETURN);
        keywords.put("super",    SUPER);
        keywords.put("this",     THIS);
        keywords.put("true",     TRUE);
        keywords.put("var",      VAR);
        keywords.put("while",    WHILE);
    }

    Scanner(String source) {
        this.source = source;
    }

    List<Token> get_tokens() {
        while (current < source.length()) {
            this.start = this.current;
            scan_token();
        }

        tokens.add(new Token(TokenType.EOF, "", null, this.line));
        return tokens;
    }

    void scan_token() {
        int c = advance();

        switch (c) {
            case '(': add_token(LEFT_PAREN); break;
            case ')': add_token(RIGHT_PAREN); break;
            case '{': add_token(LEFT_BRACE); break;
            case '}': add_token(RIGHT_BRACE); break;
            case '&': add_token(match('&') ? AND : BITWISE_AND); break;
            case '|': add_token(match('|') ? OR : BITWISE_OR); break;
            case '^': add_token(BITWISE_XOR); break;
            case '~': add_token(BITWISE_NOT); break;
            case ',': add_token(COMMA); break;
            case '.': add_token(DOT); break;
            case '-': add_token(MINUS); break;
            case '+': add_token(PLUS); break;
            case ';': add_token(SEMICOLON); break;
            case '*': add_token(STAR); break;
            case '?': add_token(QUESTION_MARK); break;
            case ':': add_token(COLON); break;
            case '=': add_token(match('=') ? EQUAL_EQUAL : EQUAL); break;
            case '!': add_token(match('=') ? BANG_EQUAL : BANG); break;
            case '<': {
                if (match('=')) {
                    add_token(LESS_EQUAL);
                } else if (match('<')) {
                    add_token(LEFT_SHIFT);
                } else {
                    add_token(LESS);
                }
            } break;
            case '>': {
                if (match('=')) {
                    add_token(GREATER_EQUAL);
                } else if (match('>')) {
                    add_token(RIGHT_SHIFT);
                } else {
                    add_token(GREATER);
                }
            } break;
            case '/':
                if (match('/')) {
                    // ignore single line comment
                    while (peek() != '\n' && !is_at_end()) current += 1;
                } else if (peek() == '*') {
                    scan_multiline_comment();
                } else {
                    add_token(SLASH);
                }
            break;

            // ignore whitespace
            case ' ':
            case '\t':
            case '\r':
            break;

            case '\n': line += 1; break;
            case '\"': scan_string_literal(); break;

            default:
                if (is_digit(c)) {
                    scan_number_literal();
                } else if (is_alpha(c)) {
                    scan_identifier();
                } else {
                    Lox.error(line, "Unexpected codepoint '" + c + "'.");
                }
            break;
        }
    }

    int peek() {
        if (current >= source.length()) return '\0';
        return source.codePointAt(current);
    }

    int peek_next() {
        if (current + 1 >= source.length()) return '\0';
        return source.codePointAt(current + 1);
    }

    int advance() {
        int c = source.codePointAt(current);
        current += 1;
        return c;
    }

    void add_token(TokenType type) {
        add_token(type, null);
    }

    void add_token(TokenType type, Object literal) {
        String text = source.substring(start, current);
        tokens.add(new Token(type, text, literal, line));
    }

    boolean match(char next) {
        if (is_at_end()) return false;
        if (source.charAt(current) != next) return false;

        current += 1;
        return true;
    }

    boolean is_digit(int c) {
        return c >= '0' && c <= '9';
    }

    boolean is_alpha(int c) {
        return (c >= 'a' && c <= 'z') ||
               (c >= 'A' && c <= 'Z') ||
               (c == '_') ||
                Character.isAlphabetic(c);
    }

    boolean is_alpha_numeric(int c) {
        return is_digit(c) || is_alpha(c);
    }

    void scan_string_literal() {
        while (peek() != '\"' && !is_at_end()) {
            if (peek() == '\n') this.line += 1;
            if (peek() == '\\') {
                current += 1;
                switch (peek()) {
                    case '0': case '1': case '2': case '3': case '4': case '5': case '6': case '7': case '8': case '9':
                    case 't': case 'b': case 'n': case 'r': case 'f': case '\'': case '\"': case '\\':
                        break;
                    default:
                        Lox.error(line, "Unrecognized escape sequence \\" + (char)peek() + ".");
                }
            }
            current += 1;
        }

        if (is_at_end()) {
            Lox.error(line, "Unterminated string.");
            return;
        }

        // To skip lexing the closing "
        current += 1;

        String value = source.substring(start + 1, current - 1);
        add_token(STRING, value);
    }
    
    void scan_number_literal() {
        while (is_digit(peek())) current += 1;

        if (peek() == '.' && is_digit(peek_next())) {
            advance();
            while (is_digit(peek())) current += 1;
        }

        double value = Double.parseDouble(source.substring(start, current));
        add_token(NUMBER, value);
    }

    void scan_identifier() {
        while (is_alpha_numeric(peek())) {
            current += 1;
        }

        String name = source.substring(start, current);
        TokenType type = keywords.get(name);
        if (type == null) type = IDENTIFIER;

        add_token(type);
    }

    void scan_multiline_comment() {
        int nested_count = 1;
        for (;; current += 1) {
            if (peek() == '\n') {
                line += 1;
                continue;
            }
            if (peek() == '/' && peek_next() == '*') {
                nested_count += 1;
            }
            if (peek() == '*' && peek_next() == '/') {
                nested_count -= 1;
            }
            if (nested_count == 0) {
                current += 2;
                break;
            }
        }
    }

    boolean is_at_end() {
        return current >= source.length();
    }
}

File: clox/src/scanner.c
#include "scanner.h"
#include <string.h>

Scanner scanner_make(const char *source_code) {
    return (Scanner){
        .start = source_code,
        .current = source_code,
        .line = 1,
        .column = 0,
    };
}

#define token_make(type) scanner_token_make(scanner, type)
static Token scanner_token_make(Scanner* scanner, TokenType type) {
    return (Token){
        .type = type,
        .start = scanner->start,
        .length = (size_t)(scanner->current - scanner->start),
        .line = scanner->line,
        .column = scanner->column,
    };
}

#define is_at_end() scanner_is_at_end(scanner)
static bool scanner_is_at_end(Scanner* scanner) {
    return *scanner->current == '\0';
}

#define token_error(message) scanner_token_error(scanner, message)
static Token scanner_token_error(Scanner* scanner, const char *message) {
    return (Token){
        .type = TOKEN_ERROR,
        .start = message,
        .length = strlen(message),
        .line = scanner->line,
        .column = scanner->column,
    };
}

#define advance() scanner_advance(scanner)
static char scanner_advance(Scanner* scanner) {
    scanner->current++;
    scanner->column++;
    return scanner->current[-1];
}

#define match(expected) scanner_match(scanner, expected)
static bool scanner_match(Scanner *scanner, char expected) {
    if (is_at_end()) return false;
    if (*scanner->current != expected) return false;

    scanner->current++;
    return true;
}

#define peek() scanner_peek(scanner)
static char scanner_peek(Scanner *scanner) {
    return *scanner->current;
}

#define peek_next() scanner_peek_next(scanner)
static char scanner_peek_next(Scanner *scanner) {
    if (is_at_end()) return '\0';
    return scanner->current[1];
}

static bool is_digit(char c) {
    return c >= '0' && c <= '9';
}

static bool is_alpha(char c) {
    return (c >= 'a' && c <= 'z') ||
           (c >= 'A' && c <= 'Z') ||
           (c == '_');
}

static bool is_alphanum(char c) {
    return is_digit(c) || is_alpha(c);
}

static void skip_whitespace(Scanner* scanner) {
    for (;;) {
        char c = peek();
        switch (c) {
            case ' ':
            case '\r':
            case '\t': advance(); break;
            case '\n': {
                scanner->line++;
                scanner->column = -1;
                advance();
            } break;
            case '/': {
                if (peek_next() == '/') {
                    while (peek() != '\n' && !is_at_end()) advance();
                } else {
                    return;
                }
            } break;
            default: return;
        }
    }
}

static Token scan_string(Scanner* scanner) {
    while (peek() != '"' && !is_at_end()) {
        if (peek() == '\n') {
            scanner->line++;
            scanner->column = -1;
        }
        advance();
    }

    if (is_at_end()) return token_error("Unterminated string.");

    advance();
    return token_make(TOKEN_STRING);
}

static Token scan_number(Scanner* scanner) {
    while (is_digit(peek())) advance();

    if (peek() == '.' && is_digit(peek_next())) {
        advance();
        while (is_digit(peek())) advance();
    }

    return token_make(TOKEN_NUMBER);
}

#define check_keyword(offset, rest, type) scanner_check_keyword(scanner, offset, sizeof(rest)-1, rest, type)
static TokenType scanner_check_keyword(Scanner* scanner, uint32_t offset, uint32_t length, const char *rest, TokenType type) {
    if (scanner->current - scanner->start == offset + length &&
        memcmp(scanner->start + offset, rest, length) == 0) {
        return type;
    }
    return TOKEN_IDENTIFIER;
}

static TokenType identifier_type(Scanner* scanner) {
    switch (scanner->start[0]) {
        case 'a': return check_keyword(1, "nd", TOKEN_AND);
        case 'c': return check_keyword(1, "lass", TOKEN_CLASS);
        case 'e': return check_keyword(1, "lse", TOKEN_ELSE);
        case 'f': {
            if (scanner->current - scanner->start > 1) {
                switch (scanner->start[1]) {
                    case 'a': return check_keyword(2, "lse", TOKEN_FALSE);
                    case 'o': return check_keyword(2, "r", TOKEN_FOR);
                    case 'u': return check_keyword(2, "n", TOKEN_FUN);
                }
            }
        } break;
        case 'i': return check_keyword(1, "f", TOKEN_IF);
        case 'n': return check_keyword(1, "il", TOKEN_NIL);
        case 'o': return check_keyword(1, "r", TOKEN_OR);
        case 'p': return check_keyword(1, "rint", TOKEN_PRINT);
        case 'r': return check_keyword(1, "eturn", TOKEN_RETURN);
        case 's': return check_keyword(1, "uper", TOKEN_SUPER);
        case 't': {
            if (scanner->current - scanner->start > 1) {
                switch (scanner->start[1]) {
                    case 'h': return check_keyword(2, "is", TOKEN_THIS);
                    case 'r': return check_keyword(2, "ue", TOKEN_TRUE);
                }
            }
        } break;
        case 'v': return check_keyword(1, "ar", TOKEN_VAR);
        case 'w': return check_keyword(1, "hile", TOKEN_WHILE);
    }
    return TOKEN_IDENTIFIER;
}

static Token scan_identifier(Scanner* scanner) {
    while (is_alphanum(peek())) advance();
    return token_make(identifier_type(scanner));
}

Token scanner_scan_token(Scanner* scanner) {
    skip_whitespace(scanner);

    scanner->start = scanner->current;

    if (is_at_end()) return token_make(TOKEN_EOF);

    char c = advance();
    if (is_digit(c)) return scan_number(scanner);
    if (is_alpha(c)) return scan_identifier(scanner);

    switch (c) {
        case '(': return token_make(TOKEN_LEFT_PAREN);
        case ')': return token_make(TOKEN_RIGHT_PAREN);
        case '{': return token_make(TOKEN_LEFT_CURLY);
        case '}': return token_make(TOKEN_RIGHT_CURLY);
        case ';': return token_make(TOKEN_SEMICOLON);
        case ',': return token_make(TOKEN_COMMA);
        case '.': return token_make(TOKEN_DOT);
        case '-': return token_make(TOKEN_MINUS);
        case '+': return token_make(TOKEN_PLUS);
        case '/': return token_make(TOKEN_SLASH);
        case '*': return token_make(TOKEN_STAR);
        case '!': return token_make(match('=') ? TOKEN_BANG_EQUAL : TOKEN_BANG);
        case '=': return token_make(match('=') ? TOKEN_EQUAL_EQUAL : TOKEN_EQUAL);
        case '<': return token_make(match('=') ? TOKEN_LESS_EQUAL : TOKEN_LESS);
        case '>': return token_make(match('=') ? TOKEN_GREATER_EQUAL : TOKEN_GREATER);
        case '"': return scan_string(scanner);
    }
    
    return token_error("Unexpected character.");
}

const char* token_type_cstring(TokenType type) {
    switch (type) {
        case TOKEN_LEFT_PAREN: return "LEFT_PAREN";
        case TOKEN_RIGHT_PAREN: return "RIGHT_PAREN";
        case TOKEN_LEFT_CURLY: return "LEFT_CURLY";
        case TOKEN_RIGHT_CURLY: return "RIGHT_CURLY";
        case TOKEN_COMMA: return "COMMA";
        case TOKEN_DOT: return "DOT";
        case TOKEN_MINUS: return "MINUS";
        case TOKEN_PLUS: return "PLUS";
        case TOKEN_SEMICOLON: return "SEMICOLON";
        case TOKEN_SLASH: return "SLASH";
        case TOKEN_STAR: return "STAR";
        case TOKEN_BANG: return "BANG";
        case TOKEN_BANG_EQUAL: return "BANG_EQUAL";
        case TOKEN_EQUAL: return "EQUAL";
        case TOKEN_EQUAL_EQUAL: return "EQUAL_EQUAL";
        case TOKEN_GREATER: return "GREATER";
        case TOKEN_GREATER_EQUAL: return "GREATER_EQUAL";
        case TOKEN_LESS: return "LESS";
        case TOKEN_LESS_EQUAL: return "LESS_EQUAL";
        case TOKEN_IDENTIFIER: return "IDENTIFIER";
        case TOKEN_STRING: return "STRING";
        case TOKEN_NUMBER: return "NUMBER";
        case TOKEN_AND: return "AND";
        case TOKEN_CLASS: return "CLASS";
        case TOKEN_ELSE: return "ELSE";
        case TOKEN_FALSE: return "FALSE";
        case TOKEN_FOR: return "FOR";
        case TOKEN_FUN: return "FUN";
        case TOKEN_IF: return "IF";
        case TOKEN_NIL: return "NIL";
        case TOKEN_OR: return "OR";
        case TOKEN_PRINT: return "PRINT";
        case TOKEN_RETURN: return "RETURN";
        case TOKEN_SUPER: return "SUPER";
        case TOKEN_THIS: return "THIS";
        case TOKEN_TRUE: return "TRUE";
        case TOKEN_VAR: return "VAR";
        case TOKEN_WHILE: return "WHILE";
        case TOKEN_ERROR: return "ERROR";
        case TOKEN_EOF: return "EOF";
        default: return NULL;
    }
}
